core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -833,8 +833,10 @@ private[spark] class BlockManager(
     logDebug("Put block %s locally took %s".format(blockId, Utils.getUsedTimeMs(startTimeMs)))
 
     // Either we're storing bytes and we asynchronously started replication, or we're storing
-    // values and need to serialize and replicate them now:
-    if (putLevel.replication > 1) {
+    // values and need to serialize and replicate them now.
+    // Should not replicate the block if its StorageLevel is StorageLevel.NONE or
+    // putting it to local is failed.
+    if (!putBlockInfo.isFailed && putLevel.replication > 1) {
```

Comment:
hi @squito @lianhuiwang,  let me know how this comment sounds ... 

"if a block is not serialized to memory locally that means we don't have enough memory. In this case we should not try to replicate the same block to remote peer even though remote peer has enough memory to serialize the block . BlockInfo status 'failed' for local block indicates that block failed to serialize locally and we can skip replicating the same block."